#  CST1510 – Programming for Data Communication & Networks Portfolio
####  By Hasti Hamilton


##  Week 1: Getting Started with Python


###  Savings Projection (3 Years)

Compute final savings after 3 years given:
- Save £10 every week for 3 years
- Year 1 bonus: +£20 each week
- Every 2 weeks: −£0.50 (sister)
Assume 52 weeks per year; use arithmetic and augmented assignment; no conditionals.

**Example Output:**
```
Final savings after 3 years: £2561.0
```

In [25]:
deduction_weeks=(52//2)  

regular_weeks = 52-deduction_weeks 

year1_regular_total = 30*(regular_weeks)

years2_3_regular_total = 10*(2*(regular_weeks))

years2_3_deduction_total = 9.5*(2*(deduction_weeks))

year1_deduction_total = 29.5*(deduction_weeks)

final_savings = year1_regular_total + years2_3_regular_total + years2_3_deduction_total + year1_deduction_total

In [26]:
print(f'Final savings after 3 years: £{final_savings}')

Final savings after 3 years: £2561.0


## Explanation of the savings Projection

In order to solve this problem, I used: variables; integer and float data types; arithmetic operations ; augmented assignment ; parentheses to control order of operations; constants (e.g., 52 weeks per year); and formatted output using print() with f-strings.


 `deduction_weeks = (52 // 2)`  
This line uses integer division (`//`) to find how many weeks per year include a sister deduction.  
Since a deduction happens every 2 weeks, 52 divided by 2 equals **26** deduction weeks.

 `regular_weeks = 52 - deduction_weeks`  
This calculates the number of weeks without any sister deduction.  
There are 52 weeks in total, minus 26 deduction weeks, leaving **26 regular weeks**.

 `year1_regular_total = 30 * (regular_weeks)`  
This finds the total savings in **Year 1** during the regular weeks.  
Each regular week in Year 1 earns £10 + £20 bonus = **£30**, so £30 × 26 = **£780**.

 `years2_3_regular_total = 10 * (2 * (regular_weeks))`  
This calculates the total savings in **Years 2 and 3** during regular weeks.  
There is no bonus in these years, so it’s £10 per week.  
Two years × 26 regular weeks = 52 weeks, and £10 × 52 = **£520**.

 `years2_3_deduction_total = 9.5 * (2 * (deduction_weeks))`  
This calculates the total savings for **Years 2 and 3** during deduction weeks.  
Each deduction week has £10 − £0.50 = **£9.50** saved.  
Two years × 26 deduction weeks = 52 weeks, and £9.5 × 52 = **£494**.

`year1_deduction_total = 29.5 * (deduction_weeks)`  
This calculates the total savings in **Year 1** during deduction weeks.  
Each deduction week has £30 − £0.50 = **£29.50** saved, so £29.5 × 26 = **£767**.

 `final_savings = year1_regular_total + years2_3_regular_total + years2_3_deduction_total + year1_deduction_total`  
This adds up all the amounts from the previous calculations to find the total savings after 3 years.  
£780 + £520 + £494 + £767 = **£2561.0** total savings.

 `print(f'Final savings after 3 years: £{final_savings}')`  
This prints the result using an **f-string**, which allows variables to be displayed inside text.  
The output will show:  
**Final savings after 3 years: £2561.0**


---

### Payroll Statement

Ask for: employee name, hours worked, hourly rate, federal tax rate, state tax rate.
Print a formatted payroll slip with gross pay, each deduction, and net pay (2 d.p.).

**Example Output:**
```
Employee: Smith
Gross pay: £97.50
Federal withholding: £19.50
State withholding: £8.78
Net pay: £69.22
```

In [15]:
name = input("Employee name: ")
hours = float(input("Hours worked: "))
rate = float(input("Hourly rate: "))
federal_tax = float(input("Federal tax rate (e.g. 0.20 for 20%): "))
state_tax = float(input("State tax rate (e.g. 0.09 for 9%): "))

gross_pay = hours * rate
federal_withholding = gross_pay * federal_tax
state_withholding = gross_pay * state_tax
net_pay = gross_pay - federal_withholding - state_withholding

print(f"\nEmployee: {name}")
print(f"Gross pay: £{gross_pay:.2f}")
print(f"Federal withholding: £{federal_withholding:.2f}")
print(f"State withholding: £{state_withholding:.2f}")
print(f"Net pay: £{net_pay:.2f}")


Employee name:  Hasti
Hours worked:  10
Hourly rate:  9.27
Federal tax rate (e.g. 0.20 for 20%):  0.20
State tax rate (e.g. 0.09 for 9%):  0.09



Employee: Hasti
Gross pay: £92.70
Federal withholding: £18.54
State withholding: £8.34
Net pay: £65.82


###  Explanation of the Payroll Program
In order to solve this problem, I used: input() for user input; type casting with float(); variables to store values; arithmetic operations ; parentheses for grouping; the escape sequence \n for a blank line; f-strings with :.2f for 2-decimal currency formatting; and print() to display results with the £ symbol.


### Input Section

**`name = input("Employee name: ")`**  
Prompts the user to type the employee’s name. The result is a **string** used only for display.

**`hours = float(input("Hours worked: "))`**  
Asks for the number of hours worked this pay period.  
`float(...)` converts the text input to a **floating point number** (for example:`37.5`).

**`rate = float(input("Hourly rate: "))`**  
Asks for the hourly wage (for example: `15.5`). Converted to **float** for arithmetic.

**`federal_tax = float(input("Federal tax rate (e.g. 0.20 for 20%): "))`**  
Requests the **rate as a decimal** (not a percent). Example: enter `0.20` for 20%. Stored as **float**.

**`state_tax = float(input("State tax rate (e.g. 0.09 for 9%): "))`**  
Same idea as above, but for state tax. Enter `0.09` for 9%.

---

### Calculations

**`gross_pay = hours * rate`**  
Gross pay is hours worked multiplied by the hourly rate.  
Example: `37.5 * 15.5 = 581.25`.

**`federal_withholding = gross_pay * federal_tax`**  
The amount taken for federal tax.  
Example: `581.25 * 0.20 = 116.25`.

**`state_withholding = gross_pay * state_tax`**  
The amount taken for state tax.  
Example: `581.25 * 0.09 = 52.3125`.

**`net_pay = gross_pay - federal_withholding - state_withholding`**  
Money the employee actually receives after both withholdings.  
Example: `581.25 - 116.25 - 52.3125 = 412.6875`.

---

### Output (Formatted)

**`print(f"\nEmployee: {name}")`**  
Prints a blank line for spacing, then shows the employee name.

**`print(f"Gross pay: £{gross_pay:.2f}")`**  
Displays gross pay rounded to **2 decimal places** using the format specifier `:.2f`.

**`print(f"Federal withholding: £{federal_withholding:.2f}")`**  
Shows federal withholding rounded to **2 d.p.**

**`print(f"State withholding: £{state_withholding:.2f}")`**  
Shows state withholding rounded to **2 d.p.**

**`print(f"Net pay: £{net_pay:.2f}")`**  
Shows net pay rounded to **2 d.p.** This matches the requirement “(2 d.p.).”

---


## Week 2: Control Structures (Selection & Iteration) + Lists

 **Student Marks Analysis**
You are working as a data analyst in a school. You have the exam marks of students stored in a list. Your task is to analyze the data.

**Requirements**:

- Store marks of students in a list.

- Use a loop to calculate the average mark.

- Find the highest and lowest mark using list functions.

- Print the number of students who passed (marks ≥ 50) and failed.

In [27]:
st = int(input("How many students? "))

total = 0
passed = 0
highest = None
lowest = None

for i in range(1,st+1):
    m = int(input(f"Mark {i}: "))
    
    total = total + m
    if m >= 50:
        passed = passed + 1
        
    if highest is None or m > highest:
        highest = m
        
    if lowest is None or m < lowest:
        lowest = m

avg = total / st
failed = st - passed

print('Average:', avg)
print('Highest:', highest)
print('Lowest:', lowest)
print('Passed:', passed)
print('Failed:', failed)

How many students?  5
Mark 1:  78
Mark 2:  69
Mark 3:  90
Mark 4:  50
Mark 5:  43


Average: 66.0
Highest: 90
Lowest: 43
Passed: 4
Failed: 1


## Student Marks Analysis 
In order to solve this problem, I used: input() for user input; type casting with int(); a list to store marks with append(); a for loop to process items; Boolean expressions like (m >= 50); comparison operators (>=, >, <, ==); control flow with if/elif/else; list functions max(), min(), len(); accumulation with += to sum marks and count passes; and formatted output with print() and f-strings.

---

### Input & Initialisation

**`st = int(input("How many students? "))`**  
Prompts for the number of students and converts it to an integer. This controls how many times we ask for a mark.

**`total = 0`**  
Running sum of all marks. We’ll use this to compute the average.

**`passed = 0`**  
Counter of students with marks ≥ 50.

**`highest = None`**  
Placeholder for the maximum mark seen so far. `None` means “not set yet”.

**`lowest = None`**  
Placeholder for the minimum mark seen so far.

---

### Loop Over Students

**`for i in range(1, st + 1):`**  
Iterates from 1 up to and including `st` to read each student’s mark once.

**`m = int(input(f"Mark {i}: "))`**  
Asks for mark *i* and converts it to an integer.

**`total = total + m`**  
Adds the current mark to the running total (for average later).

**`if m >= 50:`**  
Checks pass condition (mark ≥ 50).  
**`passed = passed + 1`**  
Increments the pass counter when the condition is true.

**`if highest is None or m > highest:`**  
On the first mark (`highest is None`) or whenever the new mark is larger, update the maximum.  
**`highest = m`**

**`if lowest is None or m < lowest:`**  
On the first mark (`lowest is None`) or whenever the new mark is smaller, update the minimum.  
**`lowest = m`**

---

### Post-Loop Calculations

**`avg = total / st`**  
Computes the average mark using the total from the loop and the student count.

**`failed = st - passed`**  
Students who failed are the remainder after subtracting the number who passed.

---

### Output

**`print('Average:', avg)`**  
Shows the average mark.

**`print('Highest:', highest)`**  
Shows the highest mark encountered.

**`print('Lowest:', lowest)`**  
Shows the lowest mark encountered.

**`print('Passed:', passed)`**  
Shows how many students scored ≥ 50.

**`print('Failed:', failed)`**  
Shows how many students scored < 50.

---

Calculate the sum of all numbers from `1` to a `given number`

Write a program to accept a number `n` from a user and calculate the sum of all numbers from 1 to a given number n


For example, if the user entered 10 the output should be 55 (1+2+3+4+5+6+7+8+9+10)

```
Expected output
Enter the value n: 10
The sum of all numbers until 10 is 55
```

In [4]:
n = int(input('Enter the value n: '))
total = 0
for i in range(1, n + 1):
    total =total+ i
print(f'The sum of all numbers until {n} is {total}')

Enter the value n:  9


The sum of all numbers until 9 is 45


## or 

In [7]:
n = int(input('Enter the value n: '))
total = n*(n + 1) // 2
print(f'The sum of all numbers until {n} is {total}')


Enter the value n:  9


The sum of all numbers until 9 is 45


##  The sum of all numbers from `1` to a `given number`
 In order to solve this problem, I used: input() for user input; type casting with int(); a running total variable; a for loop with range(1, n+1); augmented assignment (total += i); arithmetic operators with parentheses; and formatted output using print() with f-strings.

The code shows two different approaches to this problem :
1) **Loop method** 
2) **Formula method** 

---

## 1) Loop Method

 `n = int(input('Enter the value n: '))`  
Reads the upper limit from the user as text, then converts it to an integer `n`.  

`total = 0` 

Starts a running total at 0. This variable will hold the sum as we add each number .

 `for i in range(1, n + 1):`  
Creates a loop that visits all integers from **1** up to and including **n**.  
 `range(1, n + 1)` stops **before** `n + 1`, so the last value is `n`.

 `total = total + i`  
Adds the current number `i` into `total`.  
After the loop finishes, `total` equals `1 + 2 + … + n`.

`print(f'The sum of all numbers until {n} is {total}')`  
Displays the result with an f-string.  
Example output for `n = 9`: **“The sum of all numbers until 9 is 45”**.

---

## 2) Formula Method 

 `n = int(input('Enter the value n: '))`  
Reads and asks the user’s input to an integer.

 `total = n * (n + 1) // 2`  
Used a simple formula to get the total ,
 `//` is **integer division** so the result is an integer (no decimals).  
This is faster because it does not have a looping.

 `print(f'The sum of all numbers until {n} is {total}')`  
Prints the sum using an f-string.  
Example for `n = 9`: **45**.

---


 ## Week 3: List, Multidimensional Lists, Functions and Lambda Expressions


**Create a list of seven dice throws**

The throws should be randomly generated and add to the list
The list should be shuffle and print it out
Sort the list descending order and print it out
Count the occurrences and print them out
<br>


In [2]:
import random

nums = []
for i in range(7):
    nums.append(random.randint(1, 6))

random.shuffle(nums)
nums.sort(reverse=True) 

print(nums)             

for i in range(1, 7):    
    print(i, nums.count(i)) 


[5, 5, 4, 4, 4, 3, 1]
1 1
2 0
3 1
4 3
5 2
6 0


### Seven Dice Throws 
In order to solve this problem, I used: lists (creation and list comprehension); the random module with randint() and shuffle(); functions (def) with parameters; return values vs print for output; list operations sorted(reverse=True) and count(); a dictionary (via dict comprehension) to store face counts; a for-loop to display results; and formatted output with print() + f-strings.

---

#### `import random`
**Loads Python’s `random` module**  
This allows the program to generate random numbers that simulate dice rolls.

---

#### `nums = []`
**Creates an empty list** named `nums`  
This list will store the seven random dice results.

---

#### `for i in range(7):`
**Starts a loop that runs 7 times**  
Each loop represents one dice throw.

---

#### `nums.append(random.randint(1, 6))`
Each time the loop runs:
- `random.randint(1, 6)` generates a **random integer between 1 and 6**, like a real dice face.  
- `.append(...)` adds that random number to the `nums` list.

---

#### `random.shuffle(nums)`
**Randomly reorders** the items in `nums`  
This makes the list appear in a random order (it changes the list directly, not by making a new one).

---

#### `nums.sort(reverse=True)`
**Sorts** the list `nums` **from largest to smallest**  
> Note: Sorting after shuffling removes the random order, but that’s fine if the goal is to show the list in descending order.

---

#### `print(nums)`
**Displays the final list** of seven dice results (now sorted in descending order).  
**Example Output:** `[6, 5, 5, 4, 3, 2, 1]`

---

#### `for i in range(1, 7):`
**Loops through all possible dice values:** 1, 2, 3, 4, 5, and 6.  
This prepares to count how many times each value appears.

---

#### `print(i, nums.count(i))`
For each dice face `i`:
- `nums.count(i)` finds **how many times** that number appears in the list.  
- The program then prints a line showing the face and its count.  
**Example Output:** `4 2` → means the number **4 appears twice**.

---


**Rewrite the quiz or the scientific converter including meanly lambda functions.**

In [8]:
convert = {
    "c_f":  lambda x: x * 9/5 + 32,
    "f_c":  lambda x: (x - 32) * 5/9,
    "m_ft": lambda x: x * 3.28084,
    "ft_m": lambda x: x / 3.28084,
    "kg_lb": lambda x: x * 2.20462,
    "lb_kg": lambda x: x / 2.20462
}

typ = input("Enter conversion type (e.g. c_f, m_ft): ")
val = float(input("Enter value: "))

if typ in convert:
    print("Result:", convert[typ](val))
else:
    print("Invalid conversion type.")


Enter conversion type (e.g. c_f, m_ft):  c_f
Enter value:  30


Result: 86.0


## Scientific Converter Using Lambda Functions 
In order to solve this problem, I used: a dictionary to map conversion names to functions; lambda functions for each formula; user input with input(); type casting with float(); dictionary lookup and function call convert [typ] (val); conditionals (if/else) to validate the key; arithmetic operators with parentheses; and print() with f-strings for formatted output.

---

#### `convert = { ... }`
**Creates a dictionary** called `convert` that stores different conversion formulas.  
Each formula is written using a **lambda function**, which is a short, anonymous function in Python.

---

#### `"c_f": lambda x: x * 9/5 + 32,`
**Converts Celsius to Fahrenheit**  
Formula: `(°C × 9/5) + 32`

---

#### `"f_c": lambda x: (x - 32) * 5/9,`
**Converts Fahrenheit to Celsius**  
Formula: `(°F − 32) × 5/9`

---

#### `"m_ft": lambda x: x * 3.28084,`
**Converts meters to feet**  
Formula: `meters × 3.28084`

---

#### `"ft_m": lambda x: x / 3.28084,`
**Converts feet to meters**  
Formula: `feet ÷ 3.28084`

---

#### `"kg_lb": lambda x: x * 2.20462,`
**Converts kilograms to pounds**  
Formula: `kg × 2.20462`

---

#### `"lb_kg": lambda x: x / 2.20462,`
**Converts pounds to kilograms**  
Formula: `lb ÷ 2.20462`

---

#### `typ = input("Enter conversion type (e.g. c_f, m_ft): ")`
**Asks the user** to enter the desired conversion type (key from the dictionary).  
Example input: `c_f` or `kg_lb`.

---

#### `val = float(input("Enter value: "))`
**Asks the user** to enter the value they want to convert.  
The input is converted to a **float** for decimal accuracy.

---

#### `if typ in convert:`
**Checks** if the user’s entered conversion type (`typ`) exists in the `convert` dictionary.

---

#### `print("Result:", convert[typ](val))`
If the type exists:
- The program **calls the matching lambda function** using `convert[typ]`.
- It **passes the entered value** (`val`) into the function.
- **Prints the result** of the conversion.

Example output:  
`Result: 68.0`

---

#### `else:`
If the entered conversion type is **not** found in the dictionary…

---

##### `print("Invalid conversion type.")`
The program displays an **error message**:  
`Invalid conversion type.`

---

## Week 4: Strings, Special Methods, Modules and Try-Except statment


Write a program that asks the user to enter two numbers and divides them. Use try-except to catch any ZeroDivisionError and ValueError that might occur (e.g., division by zero or invalid input).

In [13]:
try:
    n1 = float(input("enter a numerator: "))
    n2 = float(input("enter a denominator: "))
    result = n1 / n2
except ValueError:
    print("Invalid input: please enter numbers.")
except ZeroDivisionError:
    print("Cannot divide by zero.")
else:
    print("Result:", result)

enter a numerator:  144
enter a denominator:  12


Result: 12.0


## Division Program with Error Handling 
In order to solve this problem, I used: input() to read two numbers; type casting with float(); variables (n1, n2, result); arithmetic division ; exception handling with try/except catching ValueError (invalid input) and ZeroDivisionError (divide by zero); the try...else clause to print only when no error occurs; and print() for output.

---

#### `try:`
**Starts a try block** — Python will “try” to run the code inside.  
If an error occurs, the program will jump to the matching `except` block instead of crashing.

---

#### `n1 = float(input("enter a numerator: "))`
**Asks the user to enter a numerator** (the top number in division).  
- The input is converted to a **float** (so it can handle decimals).  
- Example input: `144`

---

#### `n2 = float(input("enter a denominator: "))`
**Asks the user to enter a denominator** (the bottom number in division).  
- Also converted to a float.  
- Example input: `12`

---

#### `result = n1 / n2`
**Performs the division** using the values entered by the user.  
- Example calculation: `144 / 12 = 12.0`  
- If the denominator is `0`, this will cause a **ZeroDivisionError**.

---

#### `except ValueError:`
**Catches a ValueError** — happens if the user types something that’s **not a number** (like text or symbols).  

---

#### `print("Invalid input: please enter numbers.")`
If a ValueError occurs, this message is shown to the user,  
telling them to input valid numeric values only.

---

#### `except ZeroDivisionError:`
**Catches a ZeroDivisionError** — happens if the user tries to **divide by zero**.

---

#### `print("Cannot divide by zero.")`
If a ZeroDivisionError occurs, this message appears to warn the user that division by zero is not allowed.

---

#### `else:`
The `else` block **runs only if no errors occurred** in the `try` section.  
That means both inputs were valid and no division by zero happened.

---

#### `print("Result:", result)`
**Displays the final division result** if everything worked correctly.  

---

Create a function (s) counting all lower case, upper case, digits, and special symbols from a given string and print them out.
   
```python

   example_password = "P@#yn26at^&i5ve"
```

In [14]:
def c(s):
    lower =0
    upper =0
    digits =0
    special = 0
    for ch in s:
        if ch.islower():   lower  += 1
        elif ch.isupper(): upper  += 1
        elif ch.isdigit(): digits += 1
        else:              special += 1
    print("lower:", lower)
    print("upper:", upper)
    print("digits:", digits)
    print("special:", special)

example_password = "P@#yn26at^&i5ve"
c(example_password)


lower: 7
upper: 1
digits: 3
special: 4


## Count Lowercase, Uppercase, Digits, and Specials 
In order to solve this problem, I used: a function with a parameter (c(s)); a for-loop to iterate over characters; control flow with if/elif/else; string methods islower(), isupper(), and isdigit(); counters with augmented assignment (+= 1); an else branch to count special symbols; and print() to display the counts for lower, upper, digits, and special.

---

#### `def c(s):`
**Defines a function** named `c` that takes a single string argument `s` (the text to analyze).

---

#### `lower = 0`
**Counter for lowercase letters**, starts at 0.

#### `upper = 0`
**Counter for uppercase letters**, starts at 0.

#### `digits = 0`
**Counter for numeric digits (0–9)**, starts at 0.

#### `special = 0`
**Counter for special (non-alphanumeric) characters**, starts at 0.

---

#### `for ch in s:`
**Loops through each character** `ch` in the input string `s`.

---

#### `if ch.islower():    lower += 1`
If the character `ch` is a **lowercase letter** (`a–z`), **increase** the `lower` counter by 1.

---

#### `elif ch.isupper():  upper += 1`
Else, if `ch` is an **uppercase letter** (`A–Z`), **increase** the `upper` counter by 1.

---

#### `elif ch.isdigit():  digits += 1`
Else, if `ch` is a **digit** (`0–9`), **increase** the `digits` counter by 1.

---

#### `else:               special += 1`
Otherwise, the character is a **special symbol** (space or punctuation, e.g., `@ # ^ & !`), so **increase** the `special` counter by 1.

---

#### `print("lower:", lower)`
**Prints the total number of lowercase letters** found.

#### `print("upper:", upper)`
**Prints the total number of uppercase letters** found.

#### `print("digits:", digits)`
**Prints the total number of digits** found.

#### `print("special:", special)`
**Prints the total number of special characters** found.

---

#### Example usage

```python
example_password = "P@#yn26at^&i5ve"
c(example_password)

---

 ## Week 5: File Handling and Data Structures 


### Temperature Logger

Write a program that:
1. Asks the user for temperatures for 5 days
2. Writes each temperature to "temps.txt" with the day number
   Format: "Day 1: 22°C"
3. Reads the file and calculates the average temperature

<hr>

In [20]:
temp=[]
for i in range(1,6):
    t=input(f'enter the temperature for day {i} :')
    temp.append(f'{t}')
print(temp)

enter the temperature for day 1 : 20
enter the temperature for day 2 : 30
enter the temperature for day 3 : 11
enter the temperature for day 4 : 23
enter the temperature for day 5 : 10


['20', '30', '11', '23', '10']


In [21]:
t1= int(temp[0])
t2= int(temp[1])
t3=int(temp[2])
t4=int(temp[3])
t5= int(temp[4])
total = (t1+t2+t3+t4+t5)/5
print(total)

18.8


In [22]:
with open('temp.txt', 'w') as f:
    for day, t in enumerate(temp, start=1): 
        f.write(f"Day {day}: {int(t)}°C\n") 
    f.write(f'The total average of the temperatures are: {total}')

with open('temp.txt', 'r') as f:
    data = f.read()

print(data)


Day 1: 20°C
Day 2: 30°C
Day 3: 11°C
Day 4: 23°C
Day 5: 10°C
The total average of the temperatures are: 18.8


###  Temperature Logger 
In order to solve this problem, I used: input() to collect 5 values; a list to store entries with append(); a for loop with range(1, 6); list indexing (temp[0]…temp[4]); type casting with int() to convert strings to numbers; arithmetic to compute the average; the with open(...) context manager with file modes "w" and "r"; file methods write() and read(); enumerate() to label days; f-strings for formatting (including "°C" and \n newlines); and print() to display results.

---

#### `temp = []`
**Creates an empty list** named `temp` to store the user’s temperature inputs.

---

#### `for i in range(1, 6):`
**Starts a loop** that repeats 5 times — once for each day (1 to 5).

---

#### `t = input(f'enter the temperature for day {i} :')`
**Asks the user for input.**  
The `f'...'` format allows inserting the day number (`i`) directly into the text.

---

#### `temp.append(f'{t}')`
**Adds the user’s input** (`t`) to the list called `temp`.

---

#### `print(temp)`
**Displays the list of temperatures entered by the user.**  
Example output: `['20', '30', '11', '23', '10']`

---

### Converting Strings to Integers and Calculating the Average:

---

#### `t1 = int(temp[0])`
#### `t2 = int(temp[1])`
#### `t3 = int(temp[2])`
#### `t4 = int(temp[3])`
#### `t5 = int(temp[4])`
**Converts each temperature (string) into an integer** using `int()`  
so that mathematical operations can be performed.

---

#### `total = (t1 + t2 + t3 + t4 + t5) / 5`
**Adds all 5 temperatures together** and **divides by 5** to calculate the average.

---

#### `print(total)`
**Prints the calculated average.**  
Example output: `18.8`

---

### Writing Data to a File:

---

#### `with open('temp.txt', 'w') as f:`
**Opens (or creates) a file** named `temp.txt` in **write mode** (`'w'`).  
The `with` statement automatically closes the file when finished.

---

#### `for day, t in enumerate(temp, start=1):`
**Loops through** each item in the list along with its **day number**.  
`enumerate()` provides both index and value starting count at 1.

---

#### `f.write(f"Day {day}: {int(t)}°C\n")`
**Writes each day’s temperature into the file** in a neat format like:  
`Day 1: 20°C`  
The `\n` adds a **new line** after each entry.

---

#### `f.write(f'The total average of the temperatures are: {total}')`
**Adds a final line** showing the **average temperature** at the end of the file.

---

### Reading the File:

---

#### `with open('temp.txt', 'r') as f:`
**Reopens the file** `temp.txt` in **read mode** (`'r'`).

---

#### `data = f.read()`
**Reads all the contents** of the file into a variable named `data`.

---

#### `print(data)`
**Displays the file’s contents** in the output window.  

---

###  Adding Line Numbers to Python Zen

Create a Python program that reads the content from the `python_zen.txt` file and adds line numbers to each line. Each line in the output file should start with a line number, followed by a dot and a space, and then the original line from the input file. Skip numbering the first line as it is the title. Save the result into a new file called `numbered_zen.txt`.

**Steps:**
1. Open the `python_zen.txt` file for reading.
2. Skip the first line (which is the title) when numbering.
3. For each subsequent line, prepend the line number followed by a dot and a space.
4. Write the modified content to a new file called `numbered_zen.txt`.

<hr>

In [ ]:
# please download the python_zen.txt file in order for this code to work 

In [28]:
with open('python_zen.txt', 'r') as f:
    lines = f.readlines()

title = lines[0]
numbered_lines = [title]

for i, line in enumerate(lines[2:],start=1):
    numbered_lines.append(f"{i}. {line}")

with open('numbered_zen.txt', 'w') as f:
    f.writelines(numbered_lines)
    
with open('numbered_zen.txt', 'r') as f:
    data=f.read()
    
print(data)

The Zen of Python, by Tim Peters
1. Beautiful is better than ugly.
2. Explicit is better than implicit.
3. Simple is better than complex.
4. Complex is better than complicated.
5. Flat is better than nested.
6. Sparse is better than dense.
7. Readability counts.
8. Special cases aren't special enough to break the rules.
9. Although practicality beats purity.
10. Errors should never pass silently.
11. Unless explicitly silenced.
12. In the face of ambiguity, refuse the temptation to guess.
13. There should be one-- and preferably only one --obvious way to do it.
14. Although that way may not be obvious at first unless you're Dutch.
15. Now is better than never.
16. Although never is often better than *right* now.
17. If the implementation is hard to explain, it's a bad idea.
18. If the implementation is easy to explain, it may be a good idea.
19. Namespaces are one honking great idea -- let's do more of those!


### Adding Line Numbers to Python Zen 
 In order to solve this problem, I used: the with open() statement for safe file handling; file methods readlines(), writelines(), and read(); list indexing to separate the title line; a list (numbered_lines) to store modified lines; the enumerate() function with start=1 to number lines; slicing (lines[2:]) to skip the first line; f-strings for formatting line numbers (f"{i}. {line}"); and print() to display the final output from the new file.


---

#### `with open('python_zen.txt', 'r') as f:`
**Opens the file** `python_zen.txt` in **read mode** (`'r'`) and gives it the name `f` to work with inside the block.

---

#### `lines = f.readlines()`
**Reads all lines** from the file and stores them as a list named `lines`.  
Each element of this list represents one line from the text file.

---

#### `title = lines[0]`
**Takes the first line** from the list `lines` and stores it in the variable `title`.  
This is the **title of the file**, which will not be numbered.

---

#### `numbered_lines = [title]`
**Creates a new list** named `numbered_lines` that starts with the title line.

---

#### `for i, line in enumerate(lines[2:], start=1):`
**Begins a loop** that goes through all lines starting from the **third line** (index 2).  
`enumerate()` gives each line a **line number (`i`)** beginning with **1**.

---

#### `numbered_lines.append(f"{i}. {line}")`
**Adds a new formatted line** to the list `numbered_lines`.  
Each line starts with its number followed by a dot and a space.  
Example:  
`1. Beautiful is better than ugly.`

---

#### `with open('numbered_zen.txt', 'w') as f:`
**Opens (or creates)** a file called `numbered_zen.txt` in **write mode** (`'w'`).  
This file will store the new version of the text with line numbers.

---

#### `f.writelines(numbered_lines)`
**Writes all the lines** stored in `numbered_lines` to the new file.

---

#### `with open('numbered_zen.txt', 'r') as f:`
**Reopens the file** `numbered_zen.txt` in **read mode** (`'r'`) to read what was written.

---

#### `data = f.read()`
**Reads all content** from the file and stores it in the variable `data`.

---

#### `print(data)`
**Displays the final content** of the numbered Zen of Python file.  

